### PR TITLE
Refactor code in hypervisor.go

### PR DESCRIPTION
**Pull request by zedi-pramodh**
Remove dom0handle and use bool enabled to speficy if that hypervisor is enabled on this device. For kvm and kubevirt we also use base.IsHVTypeKube() to differentiate.

Cleaned up GetAvailableHypervisors()

### CODE REVIEW

File: pkg/pillar/hypervisor/hypervisor.go [hunk 1]
```diff
@@ -38,17 +38,17 @@ type Hypervisor interface {
 
 type hypervisorDesc struct {
 	constructor       func() Hypervisor
-	dom0handle        string
+	enabled           bool
```
Comment (eriknordmark) [inline pkg/pillar/hypervisor/hypervisor.go:41]: You're evaluating the existence when the Zedbox process starts and containerd etc might not be running that early.

enabled func() bool
would keep the time of the check the same.

File: pkg/pillar/hypervisor/hypervisor.go [hunk 2]
```diff
@@ -65,6 +65,13 @@ func GetHypervisor(hint string) (Hypervisor, error) {
 	}
 }
 
+// Check if the path exists.
+func isenabled(path string) bool {
```
Comment (eriknordmark) [inline pkg/pillar/hypervisor/hypervisor.go:69]: I think we already have a function for this in some package in pillar. Probably called somethingExists()

**Review by eriknordmark — CHANGES_REQUESTED** [review]

File: pkg/pillar/hypervisor/hypervisor.go (hunk 2 above)
Comment (zedi-pramodh) [reply pkg/pillar/hypervisor/hypervisor.go:69]: Let me check.

Comment (zedi-pramodh) [reply pkg/pillar/hypervisor/hypervisor.go:69]: pillar/utils/file has FileExists() will use that

File: pkg/pillar/hypervisor/hypervisor.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/hypervisor/hypervisor.go:42]: My comment still holds - why change the assumptions and behavior as part of this refactoring?

Comment (zedi-pramodh) [reply pkg/pillar/hypervisor/hypervisor.go:42]: I did not understand your comment. What behavior did I change ? May be we can chat offline.

Comment (deitch) [reply pkg/pillar/hypervisor/hypervisor.go:42]: If I understood Erik correctly, he is saying that the current behaviour just lists the file to check in the start of the program, but it does not actually check it until you call `GetAvailableHypervisors()`, which might be much later. With this change, you check it on initial startup. If one of the device files or sockets is not there yet, you will miss it.

Is this system containerd? Why would system containerd be looked at as part of the pillar hypervisor list? pillar doesn't touch it; all content and workloads are via user containerd. If it actually _is_ looking at system containerd, that is a bug.

On the other hand, I really like what you did about giving each entry its own logic. They vary so much between them, it makes more sense to keep the logic for checking each as part of it, and cleans it up a lot.

What Erik is recommending is, I believe, the following:

```go
type hypervisorDesc struct {
	constructor       func() Hypervisor
	enabled           func() bool
	hvTypeFileContent string
}

var knownHypervisors = map[string]hypervisorDesc{
	XenHypervisorName:        {constructor: newXen, enabled: func () bool {return fileutils.FileExists(nil, "/proc/xen")}, hvTypeFileContent: "xen"},
	KVMHypervisorName:        {constructor: newKvm, enabled: func() bool{fileutils.FileExists(nil, "/dev/kvm") && !base.IsHVTypeKube()}, hvTypeFileContent: "kvm"},
	KubevirtHypervisorName:   {constructor: newKubevirt, enabled: func() bool{ fileutils.FileExists(nil, "/dev/kvm") && base.IsHVTypeKube()}, hvTypeFileContent: "kubevirt"},
	ACRNHypervisorName:       {constructor: newAcrn, enabled: func() bool {fileutils.FileExists(nil, "/dev/acrn")}, hvTypeFileContent: "acrn"},
	ContainerdHypervisorName: {constructor: newContainerd, enabled: func() bool{ fileutils.FileExists(nil, "/run/containerd/containerd.sock")|},
	NullHypervisorName:       {constructor: newNull, enabled: func() bool{fileutils.DirExists(nil, "/")}},
}

func GetAvailableHypervisors() (all []string, enabled []string) {
	all = hypervisorPriority
	for _, v := range all {
		if knownHypervisors[v].enabled() {
			enabled = append(enabled, v)
		}
	}

}
```

Comment (eriknordmark) [reply pkg/pillar/hypervisor/hypervisor.go:42]: Correct @deitch

Comment (zedi-pramodh) [reply pkg/pillar/hypervisor/hypervisor.go:42]: Thanks @deitch I get it now

**Review by eriknordmark — APPROVED** [review]
LGTM

**Comment by eriknordmark** [comment]
The go tests failuire is already in master and needs to be fixed separately.

**Pull request merged**